src/tequila/apps/qcc/qcc.py

```diff
@@ -0,0 +1,394 @@
+                        ]
+                    )
+                )
+                if grad >= grad_threshold:
+                    grad_partitions.append((x_term, grad))
+
+        grad_partitions.sort(key=lambda x: x[1], reverse=True)
+        self.grad_partitions = grad_partitions
+        print("\nGradient partitions:")
+        for x, grad in grad_partitions:
+            print("{} : {}".format(x, grad))
+
+        n_dis = len(grad_partitions)
+        generators = []
+        for idx in range(n_gen):
+            part_to_sample = grad_partitions[idx % n_dis]
+            sampling_idx = idx // n_dis
+            generators.append(
+                self.get_pauli_from_x_string(
+                    part_to_sample[0], type="min_weight", sampling_index=sampling_idx
+                )
+            )
+
+        print("\nSelected QCC generators:")
+        for generator in generators:
+            print(generator)
+        self.generators.append(
+            {"g{}".format(idx): generators[idx] for idx in range(len(generators))}
+        )
+        return generators
+
+    def get_qcc_unitary(self):
+        """
+        Returns:
+            QCircuit: the QCC unitary, composed as a product of parameterized exponentiated Pauli
+                terms. The exponentiated Pauli terms are appended to the circuit following the
+                order the generators appear in `self.generators`
+        """
+
+        U_qcc = QCircuit()
+        for idx in range(len(self.generators[-1])):
+            U_qcc += tq.gates.ExpPauli(
+                angle="t{}".format(idx), paulistring=self.generators[-1]["g{}".format(idx)]
+            )
+        self.qcc_circ = U_qcc
+        self.qcircuits.append(U_qcc)
+        return U_qcc
+
+    def optimize(self, method="COBYLA", x0=None):
+        """
+        Performs the optimization of the QCC parameters.
+
+        Arguments:
+            method (str): optimizer name to be used by tq.minimize.
+            x0 (array_like): initial guess for QCC parameters. This must be a
+                vector with dimensionality equal to number of generators, i.e.
+                `self.n_gen`.
+
+        Returns:
+            the optimization result.
+        """
+
+        # can add functionality for introducing qubit mean field degrees of freedom here
+
+        if x0 is None:
+            x0 = np.full(shape=len(self.generators[-1]), fill_value=0.01)
+        else:
+            if len(x0) != len(self.generators[-1]):
+                raise ValueError(
+                    "Expected initial guess x0 to be of dimension {}, instead got {}".format(
+                        len(self.generators[-1]), len(x0)
+                    )
+                )
+
+        qcc_energy = tq.ExpectationValue(H=self.hamiltonian, U=self.ref_circ + self.qcc_circ)
+        # could add a codeblock to solve for optimal qcc energy by linear variational principle
+        # if n_gens is less than ~6?
+        result = tq.minimize(

```

Comment:
following up on other comments:
this line would then be:

result = tq.minimize(objective=qcc_energy, initial_values={...},  **self.parameters.optimizer_arguments)